src/askem_beaker/contexts/mira_model_edit/agent.py

```diff
@@ -841,7 +841,197 @@ async def change_rate_law_and_add_parameter(self,
+
+        An example of this would be "Add a new transition from S to R with the name vaccine with the rate of v. v depends on I and S"
+        Where S is the subject state, R is the outcome state, vaccine is the template_name, and v is the template_expression and ["I","S"] are the controller_names.
+
+        Always make sure that template_expression is the string representation of a mathematical expression that can be parsed by SymPy.
+        If the user provides a math expression containing the operator "^", replace it with "**".
+        If the user provides an equation, pick the right hand side expression.
+
+        Args:
+            subject_name (str): The state name that is the source of the new transition. This is the state population comes from.
+            subject_initial_value (float): The number associated with the subject state at its first step in time. If not known or not specified the default value of `1` should be used.
+            outcome_name (str): the state name that is the new transition's outputs. This is the state population moves to.
+            outcome_initial_value (float): The number associated with the output state at its first step in time. If not known or not specified the default value of `1` should be used.
+            controller_names (list[str]): The names of the controller states. These are the states that will impact the transition's rate.
+            parameter_name (str): the name of the parameter.
+            parameter_units (str): The units associated with the parameter.
+            parameter_value (float): This is a numeric value provided by the user. If not known or not specified the default value of `1` should be used.
+            parameter_description (str): The description associated with the parameter. If not known or not specified the default value of `` should be used
+            template_expression (str): The mathematical rate law for the transition.
+            template_name (str): the name of the transition.
+        """
+
+        code = agent.context.get_code("add_group_controlled_conversion_template", {
+            "subject_name": subject_name,
+            "subject_initial_value": subject_initial_value,
+            "outcome_name": outcome_name,
+            "outcome_initial_value": outcome_initial_value,
+            "controller_names": controller_names,
+            "parameter_name": parameter_name,
+            "parameter_units": parameter_units,
+            "parameter_value": parameter_value,
+            "parameter_description": parameter_description,
+            "template_expression": template_expression,
+            "template_name": template_name
+        })
+        loop.set_state(loop.STOP_SUCCESS)
+        return json.dumps(
+            {
+                "action": "code_cell",
+                "language": "python3",
+                "content": code.strip(),
+            }
+        )
+
+    @tool()
+    async def add_controlled_production_template(self,
```

Comment:
Naming consistency? Your other functions are called `add_group_controlled_xyz` and here it is missing the "group" part. 

The description is also a mix of "controlled production" and "group controlled production" and i am not sure if that matters or not when it comes to prompt injection stuff.

Reply:
great catch too much sameness and I had totally missed this